net/dns/direct.go

```diff
@@ -324,3 +324,55 @@ func (m directManager) Close() error {
 
 	return nil
 }
+
+func atomicWriteFile(fs pinholeFS, filename string, data []byte, perm os.FileMode) error {
+	var randBytes [12]byte
+	if _, err := rand.Read(randBytes[:]); err != nil {
+		return fmt.Errorf("atomicWriteFile: %w", err)
+	}
+
+	tmpName := fmt.Sprintf("%s.%x.tmp", filename, randBytes[:])
+	defer fs.Remove(tmpName)
+
+	if err := fs.WriteFile(tmpName, data, perm); err != nil {
+		return fmt.Errorf("atomicWriteFile: %w", err)
+	}
+	return fs.Rename(tmpName, filename)
+}
+
+// pinholeFS is a high-level file system abstraction designed just for use
+// by directManager, with the goal that it is easy to implement over wsl.exe.
+type pinholeFS interface {
+	Stat(name string) (isRegular bool, err error)
+	Rename(oldName, newName string) error
+	Remove(name string) error
+	ReadFile(name string) ([]byte, error)
+	WriteFile(name string, contents []byte, perm os.FileMode) error
+}
+
+// directFS is a pinholeFS implemented directly on the OS.
+type directFS struct {
+	prefix string // file path prefix; used for testing
```

Comment:
so in tests, it ends in a slash?

Reply:
no because all the passed in names are absolute. clarified in comment.